src/python/pants/rules/core/determine_source_files_test.py

```diff
@@ -0,0 +1,249 @@
+        adaptors_with_origins: Iterable[TargetAdaptorWithOrigin],
+        *,
+        strip_source_roots: bool = False,
+    ) -> List[str]:
+        request = AllSourceFilesRequest(
+            (adaptor_with_origin.adaptor for adaptor_with_origin in adaptors_with_origins),
+            strip_source_roots=strip_source_roots,
+        )
+        result = self.request_single_product(
+            SourceFiles, Params(request, create_options_bootstrapper())
+        )
+        return sorted(result.snapshot.files)
+
+    def get_specified_source_files(
+        self,
+        adaptors_with_origins: Iterable[TargetAdaptorWithOrigin],
+        *,
+        strip_source_roots: bool = False,
+    ) -> List[str]:
+        request = SpecifiedSourceFilesRequest(
+            adaptors_with_origins, strip_source_roots=strip_source_roots,
+        )
+        result = self.request_single_product(
+            SourceFiles, Params(request, create_options_bootstrapper())
+        )
+        return sorted(result.snapshot.files)
+
+    def test_address_specs(self) -> None:
+        target1 = self.mock_target(
+            self.SOURCES1, origin=SingleAddress(directory=self.SOURCES1.source_root, name="lib")
+        )
+        target2 = self.mock_target(
+            self.SOURCES2, origin=SiblingAddresses(self.SOURCES2.source_root)
+        )
+        target3 = self.mock_target(
+            self.SOURCES3, origin=DescendantAddresses(self.SOURCES3.source_root)
+        )
+        target4 = self.mock_target(
+            self.SOURCES1, origin=AscendantAddresses(self.SOURCES1.source_root)
+        )
+
+        def assert_all_source_files_resolved(
+            target: TargetAdaptorWithOrigin, sources: TargetSources
+        ) -> None:
+            expected = sources.source_file_absolute_paths
+            assert self.get_all_source_files([target]) == expected
+            assert self.get_specified_source_files([target]) == expected
+
+        assert_all_source_files_resolved(target1, self.SOURCES1)
+        assert_all_source_files_resolved(target2, self.SOURCES2)
+        assert_all_source_files_resolved(target3, self.SOURCES3)
+        assert_all_source_files_resolved(target4, self.SOURCES1)
+        # NB: target1 and target4 refer to the same files. We should be able to handle this
+        # gracefully.
+        combined_targets = [target1, target2, target3, target4]
+        combined_expected = sorted(
+            [
+                *self.SOURCES1.source_file_absolute_paths,
+                *self.SOURCES2.source_file_absolute_paths,
+                *self.SOURCES3.source_file_absolute_paths,
+            ]
+        )
+        assert self.get_all_source_files(combined_targets) == combined_expected
+        assert self.get_specified_source_files(combined_targets) == combined_expected
+
+    def test_filesystem_specs(self) -> None:
+        # Literal file arg.
+        target1_all_sources = self.SOURCES1.source_file_absolute_paths
+        target1_slice = slice(0, 1)
```

Comment:
This is the same thing as `[0:1]`. You can't write `[0:1]` unless it's directly applied to a sequence; this is how you store a `slice` in a variable.